File: 2021/java/1642.furthest-building-you-can-reach.java
class Solution {
    // idea is: using min heap, 
    // 1. if next height > prev height (diff) > 0, use ladder
    // 2. if all ladders used, pop out the smallest diff, start using bricks
    // 3. if found bricks < 0, then cannot reach further, return
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        int len = heights.length;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int i = 0; i < len - 1; i++) {
            int diff = heights[i + 1] - heights[i];
            if (diff > 0) {
                pq.offer(diff);
            }
            if (pq.size() > ladders) {
                bricks -= pq.poll();
            }
            if (bricks < 0) return i;
        }
        return len - 1;
    }
}

File: 2021/python/1642.furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        if not heights or len(heights) == 0:
            return 0
        min_heap = []
        for i in range(len(heights) - 1):
            diff = heights[i + 1] - heights[i]
            if diff > 0:
                heapq.heappush(min_heap, diff)
            if len(min_heap) > ladders:
                bricks -= heapq.heappop(min_heap)
            if bricks < 0:
                return i
        return len(heights) - 1
